`logical_line` for call with f-string can result in invalid ast under Python 3.12

### how did you install flake8?

```console
$ pip install flake8
```


### unmodified output of `flake8 --bug-report`

```json
{}
```


### describe the problem

#### what I expected to happen

Some of the plugins in [`hacking`](https://pypi.org/project/hacking) use a combo of `ast.parse` and `logical_line` to generate an AST for an individual line that is then fed into an `ast.NodeVisitor` subclass. These work just fine on Python 3.11 and earlier, but I've noticed they fail under specific circumstances on Python 3.12. I've included a minimal reproducer below. There's a chance I am "holding it wrong" but I'd like to confirm this first.

#### sample code

`test.py`

```python
import unittest


class TestFoo(unittest.TestCase):

    def test_foo(self):
        response_key = 'foo'
        response_val = 'bar'
        body_output = ''
        self.assertEqual(
            f'{{"{response_key}": "{response_val}"}}', body_output
        )
```

`checks.py`:

```
class NoneArgChecker(ast.NodeVisitor):
    def __init__(self, func_name, num_args=2):
        self.func_name = func_name
        self.num_args = num_args
        self.none_found = False

    def visit_Call(self, node):
        # snipped
        ...


def foo(logical_line, noqa):
    if noqa:
        return

    for func_name in (
        'assertEqual', 'assertIs', 'assertNotEqual', 'assertIsNot'
    ):
        try:
            start = logical_line.index('.%s(' % func_name) + 1
        except ValueError:
            continue
        checker = NoneArgChecker(func_name)
        # print(logical_line)
        checker.visit(ast.parse(logical_line))
        continue
        if checker.none_found:
            yield start, "H203: Use assertIs(Not)None to check for None"
```

`tox.ini`:

```ini
[flake8:local-plugins]
extension =
    X123 = checks:foo
paths = .
```

#### commands ran

```console
$ flake8 test.py
wow.py:1:23: E999 SyntaxError: invalid syntax. Perhaps you forgot a comma?
```

If I comment out the `print` statement, I see different results for Python 3.12 compared Pythons 3.10 and 3.11. Under 3.12:

```
self.assertEqual(f'x{x{response_key}xxxx{response_val}xx}', body_output)
```

Under 3.10 and 3.11:

```
self.assertEqual(f'xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx', body_output)
```

### additional information

The E999 error is associated with line 1 of the file, despite the error actually coming from a plugin. This led me on a wild goose chase as I tried to figure out why flake8 thought my file was invalid Python but Python itself did not. I suspect this might also be user error and I should be handling the potential exception from `ast.parse` in my plugin but again, I just wanted to confirm that this was expected practice.

--- a/src/flake8/processor.py
+++ b/src/flake8/processor.py
@@ -203,7 +203,13 @@ class FileProcessor:
             if token_type == tokenize.STRING:
                 text = mutate_string(text)
             elif token_type == FSTRING_MIDDLE:  # pragma: >=3.12 cover
-                text = "x" * len(text)
+                # A curly brace in an FSTRING_MIDDLE token must be an escaped
+                # curly brace. Both 'text' and 'end' will account for the
+                # escaped version of the token (i.e. a single brace) rather
+                # than the raw double brace version, so we must counteract this
+                brace_offset = text.count("{") + text.count("}")
+                text = "x" * (len(text) + brace_offset)
+                end = (end[0], end[1] + brace_offset)
             if previous_row:
                 (start_row, start_column) = start
                 if previous_row != start_row:

--- a/tests/integration/test_plugins.py
+++ b/tests/integration/test_plugins.py
@@ -1,6 +1,8 @@
 """Integration tests for plugin loading."""
 from __future__ import annotations
 
+import sys
+
 import pytest
 
 from flake8.main.cli import main
@@ -258,6 +260,39 @@ f'hello world'
 
     expected = """\
 t.py:1:1: T001 "f'xxxxxxxxxxx'"
+"""
+    out, err = capsys.readouterr()
+    assert out == expected
+
+
+def test_escaping_of_fstrings_in_string_redacter(tmpdir, capsys):
+    cfg_s = f"""\
+[flake8]
+extend-ignore = F
+[flake8:local-plugins]
+extension =
+    T = {yields_logical_line.__module__}:{yields_logical_line.__name__}
+"""
+
+    cfg = tmpdir.join("tox.ini")
+    cfg.write(cfg_s)
+
+    src = """\
+f'{{"{hello}": "{world}"}}'
+"""
+    t_py = tmpdir.join("t.py")
+    t_py.write_binary(src.encode())
+
+    with tmpdir.as_cwd():
+        assert main(("t.py", "--config", str(cfg))) == 1
+
+    if sys.version_info >= (3, 12):  # pragma: >=3.12 cover
+        expected = """\
+t.py:1:1: T001 "f'xxx{hello}xxxx{world}xxx'"
+"""
+    else:  # pragma: <3.12 cover
+        expected = """\
+t.py:1:1: T001 "f'xxxxxxxxxxxxxxxxxxxxxxxx'"
 """
     out, err = capsys.readouterr()
     assert out == expected

